don't use hardcoded paths

phreaker0 opened the pull request:
This PR removes the hard coded paths for all binaries and executes them bare, relying on a correct set $PATH. For compatibility reasons and CRON all previous used binary locations are added PATH environment variable. So this should only break the execution on the remote hosts if the PATH isn't set correctly there?

@jimsalterjrs Is this enough for you or should I also figure out how to adapt the PATH on the remote as well (i wasn't very successful with first attempts)

Fixes #21 
Fixes #85

redmop commented:
So what if we make the executables a variable. Like $zpool="/sbin/zpool" or some such thing (don't actually use the variable name, I just could not think of anything better atm).

phreaker0 commented:
@jimsalterjrs I just checked and i didn't rolled it out yet on my production systems, only on my machine. Will deploy it soon to ubuntu/debian machines and will report back after a while.

phreaker0 commented:
@redmop They currently are in a variable, but if the target host uses different locations for the binaries it won't work. This case should be fixed by this PR, but can also potentially break cases where the remote hasn't setuped a correct $PATH environment variable.

redmop commented:
The idea behind that is it would be easier to alter the binary path if we put it in a variable. Then if we tried to find it, we just change the variable and we're done.

phreaker0 commented on syncoid:
```diff
@@ -48,17 +48,20 @@ my $debug = $args{'debug'};
+# currently using which to check for command existence because we aren't depending on perl
 # being present on remote machines.
-my $lscmd = '/bin/ls';
+my $whichcmd = 'which';
```
will do, thanks!

tomhoover commented:
You should also remove the hardcoded paths from lines 676-720 for the various compression methods.

phreaker0 commented:
@tomhoover thx

tomhoover commented:
I've tested on both MacOS (Mojave) and Debian 9--works well on both.

jimsalterjrs commented:
@phreaker0 did you say you've been using this in prod now without an explicitly set PATH in crontab? If so, on what distros?

phreaker0 commented:
@phreaker0 I only run sanoid via systemd timers. But I switched my workstation to use cron and removed the default PATH set by the distribution (ubuntu 18.04), will report back next week.

phreaker0 commented:
I apparently never reported back, it worked for me on ubuntu 18.04 with cron. I just also tested syncoid from 18.04 to FreeBSD 11.2 with that patch and it also worked.

jimsalterjrs commented:
@phreaker0 will you rebase, please?

phreaker0 commented:
@jimsalterjrs done

jimsalterjrs commented:
TY!

Now if only there was a way to fix the issue with the shebang still needing to point to the correct location of Perl...

phreaker0 commented:
@jimsalterjrs what about `#!/usr/bin/env perl`?

jimsalterjrs commented:
Last time I tested, that did not work properly in a default cron environment on Ubuntu or FreeBSD.

phreaker0 commented:
true, if $PATH is not populated this won't work. So I guess there is no easy way then.

redmop commented:
We can suggest that they set the PATH in crontab.

phreaker0 commented:
@redmop this would be the proper solution, but would break some existing setups
